Add StrokeAlign to Border

bernaferrari opened the pull request:
Fix https://github.com/flutter/flutter/issues/88621.

This is not a breaking change, however, we probably want BorderSide.toString() to include the new StrokeAlign property, which kind of makes it a breaking change. I had to update 20 tests. We could live without this, it works the same without toString(), but I think it was the correct thing to do.

Left, Figma. Right, Flutter:
![image](https://user-images.githubusercontent.com/351125/163880493-7fe7eaa5-6b62-4cc0-87f8-1fd5e510c968.png)

The change is relatively simple and allows nice things in the future, such as [rings](https://tailwindcss.com/docs/ring-color) which are popular in Tailwind.

Important to know:
- Lerp is HARD. There is no 'simple way' to lerp from inside to outside or center to outside. But, all strokeAlign values share a similar state: zero. If the strokeAlign changes, from t = 0 to t = 0.49, a.width becomes 0, then from t = 0.50 to t = 1, 0 becomes b.width. This makes the animation fluid (however kind of weird, as there are two different speeds, so it doesn't feel linear). We could also get the t where they would linearly intersect and apply the same strategy there, so it is REALLY linear, instead of fast until 0.49 and slow until 1.

https://user-images.githubusercontent.com/351125/163881225-5e3be3f3-b926-4448-8e67-1dbaba3e12f6.mov

- Border(left: BorderSide(strokeAlign: StrokeAlign.outside)) is not supported because its implementation is different (and we probably don't want this to happen, like BorderRadius is also forbidden).

- Border and all OutlinedBorders were updated so it gets half the padding on center and no padding at all at outside.

Code sample to play with my changes:
```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData.dark(),
      home: const Scaffold(body: MyHomePage()),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

List<Widget> spaceRow(double gap, Iterable<Widget> children) => children
    .expand((item) sync* {
      yield SizedBox(width: gap);
      yield item;
    })
    .skip(1)
    .toList();

List<Widget> spaceColumn(double gap, Iterable<Widget> children) => children
    .expand((item) sync* {
      yield SizedBox(height: gap);
      yield item;
    })
    .skip(1)
    .toList();

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    const double width = 40;
    final Color borderColor = Colors.green.withOpacity(0.3);
    const debugStrokeAlign = StrokeAlign.outside;

    final circleLerp = CircleBorder(
      side: BorderSide(
        color: Colors.red.withOpacity(0.5),
        width: 15,
        strokeAlign: debugStrokeAlign,
      ),
    );

    final stadiumLerp = StadiumBorder(
      side: BorderSide(
        color: Colors.red.withOpacity(0.5),
        width: 15,
        strokeAlign: debugStrokeAlign,
      ),
    );

    final roundedRectangleLerp = RoundedRectangleBorder(
      side: BorderSide(
        color: Colors.red.withOpacity(0.5),
        width: 15,
        strokeAlign: debugStrokeAlign,
      ),
    );

    ShapeBorder stadiumToCircleLerp(double t) =>
        stadiumLerp.lerpTo(circleLerp, t)!;

    ShapeBorder stadiumToRoundedLerp(double t) =>
        stadiumLerp.lerpTo(roundedRectangleLerp, t)!;

    ShapeBorder roundedToCircleLerp(double t) =>
        roundedRectangleLerp.lerpTo(circleLerp, t)!;

    return ListView(
      padding: const EdgeInsets.all(50),
      children: [
        const Text(
          "Material Borders",
          textAlign: TextAlign.center,
        ),
        SizedBox(height: 16),
        Row(
          crossAxisAlignment: CrossAxisAlignment.center,
          mainAxisAlignment: MainAxisAlignment.center,
          children: spaceRow(
            width * 2,
            [
              for (StrokeAlign align in StrokeAlign.values)
                Column(
                  children: spaceColumn(
                    16,
                    [
                      Text(align.name),
                      const SizedBox(height: 16),
                      SizedBox(
                        width: 60,
                        height: 40,
                        child: Material(
                          color: Colors.black,
                          child: Center(
                            child: Container(
                              width: 40,
                              height: 2,
                              color: Colors.red,
                            ),
                          ),
                          shape: CircleBorder(
                            side: BorderSide(
                              color: borderColor,
                              width: 10,
                              strokeAlign: align,
                            ),
                          ),
                        ),
                      ),
                      const SizedBox(height: 16),
                      Material(
                        color: Colors.black,
                        child: SizedBox(
                          width: 100,
                          height: 40,
                          child: Center(
                            child: Container(
                              width: 100,
                              height: 2,
                              color: Colors.red,
                            ),
                          ),
                        ),
                        shape: RoundedRectangleBorder(
                          side: BorderSide(
                            color: borderColor,
                            width: 20,
                            strokeAlign: align,
                          ),
                        ),
                      ),
                      const SizedBox(height: 40),
                      Material(
                        color: Colors.black,
                        child: SizedBox(
                          width: 100,
                          height: 20,
                          child: Center(
                            child: Container(
                              width: 100,
                              height: 2,
                              color: Colors.red,
                            ),
                          ),
                        ),
                        shape: StadiumBorder(
                          side: BorderSide(
                            color: borderColor,
                            width: 5,
                            strokeAlign: align,
                          ),
                        ),
                      ),
                      const SizedBox(height: 16),
                      Material(
                        color: Colors.black,
                        child: SizedBox(
                          width: 100,
                          height: 40,
                          child: Center(
                            child: Container(
                              width: 100,
                              height: 2,
                              color: Colors.red,
                            ),
                          ),
                        ),
                        shape: BeveledRectangleBorder(
                          borderRadius: BorderRadius.circular(10),
                          side: BorderSide(
                            color: borderColor,
                            width: 5,
                            strokeAlign: align,
                          ),
                        ),
                      ),
                      const SizedBox(height: 16),
                      Material(
                        color: Colors.black,
                        child: SizedBox(
                          width: 100,
                          height: 40,
                          child: Center(
                            child: Container(
                              width: 100,
                              height: 2,
                              color: Colors.red,
                            ),
                          ),
                        ),
                        shape: ContinuousRectangleBorder(
                          side: BorderSide(
                            color: Colors.red.withOpacity(0.5),
                            width: 15,
                            strokeAlign: align,
                          ),
                          borderRadius: BorderRadius.circular(30),
                        ),
                      ),
                      // const SizedBox(height: 16),
                      // OutlinedButton(
                      //   onPressed: () {},
                      //   child: const Text("helloo"),
                      //   style: OutlinedButton.styleFrom(
                      //     backgroundColor: Colors.blue,
                      //     side: BorderSide(
                      //       color: Colors.red.withOpacity(0.5),
                      //       width: 5,
                      //       strokeAlign: align,
                      //     ),
                      //   ),
                      // ),
                    ],
                  ),
                ),
            ],
          ),
        ),
        const SizedBox(height: 16),
        const Text(
          "Lerp StadiumBorder -> RoundedRectangleBorder",
          textAlign: TextAlign.center,
        ),
        const SizedBox(height: 16),
        Row(
          children: spaceRow(32, [
            Container(
              width: 100,
              height: 60,
              decoration: ShapeDecoration(
                shape: stadiumToRoundedLerp(0.01),
              ),
            ),
            Container(
              width: 100,
              height: 60,
              decoration: ShapeDecoration(
                color: Colors.yellow.withOpacity(0.5),
                shape: stadiumToRoundedLerp(0.5),
              ),
            ),
            Container(
              width: 100,
              height: 60,
              decoration: ShapeDecoration(
                color: Colors.blue,
                shape: stadiumToRoundedLerp(0.99),
              ),
            ),
            Container(
              width: 100,
              height: 60,
              decoration: ShapeDecoration(
                color: Colors.purple,
                shape: stadiumToRoundedLerp(1.0),
              ),
            ),
            Container(
              width: 100,
              height: 60,
              decoration: ShapeDecoration(shape: roundedRectangleLerp),
            ),
          ]),
        ),
        const SizedBox(height: 16),
        const Text(
          "Lerp StadiumBorder -> CircleBorder",
          textAlign: TextAlign.center,
        ),
        const SizedBox(height: 16),
        Row(
          children: spaceRow(32, [
            Container(
              width: 100,
              height: 60,
              decoration: ShapeDecoration(
                shape: stadiumToCircleLerp(0.01),
              ),
            ),
            Container(
              width: 100,
              height: 60,
              decoration: ShapeDecoration(
                color: Colors.yellow.withOpacity(0.5),
                shape: stadiumToCircleLerp(0.5),
              ),
            ),
            Container(
              width: 100,
              height: 60,
              decoration: ShapeDecoration(
                color: Colors.blue,
                shape: stadiumToCircleLerp(0.99),
              ),
            ),
            Container(
              width: 100,
              height: 60,
              decoration: ShapeDecoration(
                color: Colors.purple,
                shape: stadiumToCircleLerp(1.0),
              ),
            ),
            Container(
              width: 100,
              height: 60,
              decoration: ShapeDecoration(shape: circleLerp),
            ),
          ]),
        ),
        const SizedBox(height: 16),
        const Text(
          "Lerp RoundedRectangleBorder -> CircleBorder",
          textAlign: TextAlign.center,
        ),
        const SizedBox(height: 16),
        Row(
          children: spaceRow(32, [
            Container(
              width: 100,
              height: 100,
              decoration: ShapeDecoration(
                shape: roundedToCircleLerp(0.01),
              ),
            ),
            Container(
              width: 100,
              height: 100,
              decoration: ShapeDecoration(
                color: Colors.yellow.withOpacity(0.5),
                shape: roundedToCircleLerp(0.5),
              ),
            ),
            Container(
              width: 100,
              height: 100,
              decoration: ShapeDecoration(
                color: Colors.blue,
                shape: roundedToCircleLerp(0.99),
              ),
            ),
            Container(
              width: 100,
              height: 100,
              decoration: ShapeDecoration(
                color: Colors.purple,
                shape: roundedToCircleLerp(1.0),
              ),
            ),
            Container(
              width: 100,
              height: 100,
              decoration: ShapeDecoration(
                shape: circleLerp,
              ),
            ),
          ]),
        ),
        const SizedBox(height: 48),
        const Text(
          "Transitions",
          textAlign: TextAlign.center,
        ),
        const SizedBox(height: 20),
        const AnimatedTransitions(),
        const SizedBox(height: 48),
        const Text(
          "Lerp",
          textAlign: TextAlign.center,
        ),
        const SizedBox(height: 20),
        const Lerpness(),
        const SizedBox(height: 20),
        const Text(
          "Container",
          textAlign: TextAlign.center,
        ),
        const SizedBox(height: 40),
        Row(
          crossAxisAlignment: CrossAxisAlignment.center,
          mainAxisAlignment: MainAxisAlignment.center,
          children: spaceRow(
            width * 2,
            [
              for (StrokeAlign align in StrokeAlign.values)
                Container(
                  width: 100,
                  height: 100,
                  decoration: BoxDecoration(
                    color: Colors.red,
                    borderRadius: BorderRadius.circular(20),
                    border: Border.all(
                      color: Colors.black.withOpacity(0.75),
                      width: width,
                      strokeAlign: align,
                    ),
                  ),
                  child: Center(
                      child: Container(
                          width: 100, height: 100, color: Colors.blue)),
                ),
            ],
          ),
        ),
        const SizedBox(height: width * 2),
        Row(
          crossAxisAlignment: CrossAxisAlignment.center,
          mainAxisAlignment: MainAxisAlignment.center,
          children: spaceRow(
            width * 2,
            [
              for (StrokeAlign align in StrokeAlign.values)
                Container(
                  width: 100,
                  height: 100,
                  decoration: BoxDecoration(
                    color: Colors.green,
                    border: Border.all(
                      color: Colors.black.withOpacity(0.75),
                      width: width,
                      strokeAlign: align,
                    ),
                  ),
                ),
            ],
          ),
        ),
        const SizedBox(height: width * 2),
        Row(
          crossAxisAlignment: CrossAxisAlignment.center,
          mainAxisAlignment: MainAxisAlignment.center,
          children: spaceRow(
            width * 2,
            [
              for (StrokeAlign align in StrokeAlign.values)
                Container(
                  width: 100,
                  height: 100,
                  decoration: BoxDecoration(
                    color: Colors.blue,
                    shape: BoxShape.circle,
                    border: Border.all(
                      color: Colors.black.withOpacity(0.75),
                      width: width,
                      strokeAlign: align,
                    ),
                  ),
                ),
            ],
          ),
        ),
      ],
    );
  }
}

class AnimatedTransitions extends StatefulWidget {
  const AnimatedTransitions({Key? key}) : super(key: key);

  @override
  _AnimatedTransitionsState createState() => _AnimatedTransitionsState();
}

class _AnimatedTransitionsState extends State<AnimatedTransitions>
    with TickerProviderStateMixin {
  late AnimationController _controller;
  late Tween<double> tween;

  @override
  initState() {
    super.initState();
    _controller = AnimationController(
      duration: const Duration(seconds: 3),
      vsync: this,
    )..repeat();
  }

  @override
  dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
        animation:
            CurvedAnimation(parent: _controller, curve: Curves.slowMiddle),
        builder: (context, child) {
          return Center(
            child: Row(
              mainAxisAlignment: MainAxisAlignment.center,
              children: <Widget>[
                Material(
                  color: Colors.black,
                  child: SizedBox(
                    width: 100,
                    height: 40,
                    child: Center(
                      child: Container(
                        width: 100,
                        height: 2,
                        color: Colors.red,
                      ),
                    ),
                  ),
                  shape: _controller.value > 0.5
                      ? RoundedRectangleBorder(
                          side: BorderSide(
                            color: Colors.red.withOpacity(0.5),
                            width: 15,
                            strokeAlign: StrokeAlign.center,
                          ),
                          borderRadius: BorderRadius.circular(30),
                        )
                      : CircleBorder(
                          side: BorderSide(
                            color: Colors.red.withOpacity(0.5),
                            width: 15,
                            strokeAlign: StrokeAlign.outside,
                          ),
                        ),
                ),
                const SizedBox(width: 20),
                Material(
                  color: Colors.black,
                  child: SizedBox(
                    width: 100,
                    height: 40,
                    child: Center(
                      child: Container(
                        width: 100,
                        height: 2,
                        color: Colors.red,
                      ),
                    ),
                  ),
                  shape: _controller.value > 0.5
                      ? StadiumBorder(
                          side: BorderSide(
                            color: Colors.red.withOpacity(0.5),
                            width: 15,
                            strokeAlign: StrokeAlign.outside,
                          ),
                        )
                      : CircleBorder(
                          side: BorderSide(
                            color: Colors.red.withOpacity(0.5),
                            width: 15,
                            strokeAlign: StrokeAlign.outside,
                          ),
                        ),
                ),
                const SizedBox(width: 20),
                Material(
                  color: Colors.black,
                  child: SizedBox(
                    width: 100,
                    height: 40,
                    child: Center(
                      child: Container(
                        width: 100,
                        height: 2,
                        color: Colors.red,
                      ),
                    ),
                  ),
                  shape: _controller.value > 0.5
                      ? StadiumBorder(
                          side: BorderSide(
                            color: Colors.red.withOpacity(0.5),
                            width: 15,
                            strokeAlign: StrokeAlign.outside,
                          ),
                        )
                      : RoundedRectangleBorder(
                          side: BorderSide(
                            color: Colors.red.withOpacity(0.5),
                            width: 15,
                            strokeAlign: StrokeAlign.outside,
                          ),
                        ),
                ),
              ],
            ),
          );
        });
  }
}

class Lerpness extends StatefulWidget {
  const Lerpness({Key? key}) : super(key: key);

  @override
  State<Lerpness> createState() => _LerpnessState();
}

class _LerpnessState extends State<Lerpness> {
  bool direction = true;

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: () {
        setState(() {
          direction = !direction;
        });
      },
      child: AnimatedContainer(
        width: 100,
        height: 100,
        decoration: BoxDecoration(
          color: Colors.red,
          borderRadius: BorderRadius.circular(20),
          border: Border.all(
            color: Colors.black.withOpacity(0.75),
            width: direction ? 20 : 10,
            strokeAlign: direction ? StrokeAlign.inside : StrokeAlign.outside,
          ),
        ),
        duration: const Duration(seconds: 2),
      ),
    );
  }
}
```

bernaferrari committed:
Add StrokeAlign

bernaferrari committed:
Add lerp.

bernaferrari committed:
Fix toString() tests.

bernaferrari committed:
Add tests.

bernaferrari committed:
Fix analyzer.

bernaferrari committed:
Add dimensions support.

bernaferrari committed:
Improve dimensions.

bernaferrari committed:
Use switch instead of if/else.

bernaferrari committed:
Fix analyzer.

bernaferrari committed:
Fix test.

bernaferrari committed:
Fix docs.

bernaferrari committed:
Fix tiny thing.

bernaferrari committed:
Add RoundedRectangleBorder and CircleBorder support.

bernaferrari committed:
Fix space.

bernaferrari committed:
Fix analyzer.

bernaferrari committed:
Refactor OutsideBorders (Rounded, Stadium, Beveled, Continous).

bernaferrari committed:
Update tests.

bernaferrari commented:
I just committed big changes refactoring my PR (it was already ready for review, but while making another PR I learned a few tricks and came back here):

- Add StrokeAlign to basically all OutlinedBorders. RoundedRectangleBorder, ContinousRectangleBorder, StadiumBorder and BeveledRectangleBorder.
- I changed RoundedRectangleBorder to draw using a `rrect` instead of `drrect`. Technically this is a "breaking change" because I had to update a test. One parameter is better than two parameters, code with the StrokeAlign gets cleaner and overall all borders work the same now. If this change isn't welcome, I can revert.
```dart
// Before
..drrect(
    outer: RRect.fromRectAndRadius(rect, const Radius.circular(100.0)),
    inner: RRect.fromRectAndRadius(rect.deflate(10.0), const Radius.circular(90.0)),
)
// After
rrect(
    rrect: RRect.fromRectAndRadius(rect, const Radius.circular(100.0)).deflate(5.0),
)
```
- I had changed `dimensions` for `center` and `outside`. Now, I also changed `innerPath` depending on StrokeAlign. The "issue" is, `outerPath` must not be changed or stranger things happen (like it affects the widget width). So, the innerPath gets smaller when `StrokeAlign.center` and is the same as outerPath for `StrokeAlign.outside`. The difference on `StrokeAlign.outside` should be in the drawing only, the Border shouldn't occupy space. I think this is fine, but I wanted to document this behavior. I didn't forget outerPath implementations :P.
-  There is a lot of duplicated code between the borders now. Should we keep it that way or maybe we could change `OutlinedBorder` to include a default `dimensions` and two helper functions for `getInnerPath` and `paint`, that every inherited class could use? What do you think? I tried to be more conservative (the changes are already kind of big and hard to test/see/understand).
- I updated the sample code, basically showing how everything becomes now: 
<img width="943" alt="image" src="https://user-images.githubusercontent.com/351125/168496968-bf690e02-2e6c-4b99-a0d0-29b1ebe9c3eb.png">
- In doing so, I also fixed how CircleBorder lerps to RoundedRectangleBorder and so on. You can see on the sample, it is perfect.
- There are two outlier behaviors:
1. BeveledRectangleBorder by default has a few pixels set to border outside, and when you set to `StrokeAlign.outside`, it has a few pixels to the inside. I think this is fine. I'm not breaking its behavior and it is consistent.
2. ContinousRectangleBorder by default works as `StrokeAlign.center` (while everything else in the galaxy works as `StrokeAlign.inside`). As 'inside' doesn't make a lot of sense there, I made inside = center, only outside is different (you can also see on the sample).
- Tests are hard. RoundedRectangleBorder has a few goldens, the other ones don't. I tried to test based on some points I knew would exist/not exist, but testing paint is a pain.

bernaferrari committed:
Fix analyzer.

bernaferrari committed:
Fix roundedRectangleBorder.

bernaferrari committed:
Revert continuous_rectangle_border changes.

gspencergoog commented on packages/flutter/lib/src/painting/borders.dart:
```diff
@@ -21,6 +21,27 @@ enum BorderStyle {
   // if you add more, think about how they will lerp
 }
 
+/// The direction of the stroke to be painted for a [BorderSide] in a [Border].
```
```suggestion
/// The relative position of the stroke on a [BorderSide] in a [Border].
```

bernaferrari replied:
question: do I leave it that way or write `in a Border or OutlinedBorder`? Technically it belongs to the two.

gspencergoog replied:
Sure, add in the `OutlinedBorder`.

gspencergoog commented on packages/flutter/lib/src/painting/borders.dart:
```diff
@@ -21,6 +21,27 @@ enum BorderStyle {
+/// - [center] provides padding with half [BorderSide.width].
+/// - [outside] provides zero padding, as stroke is drawn entirely outside.
+enum StrokeAlign {
+  /// Border is drawn inside.
```
```suggestion
  /// The border is drawn on the inside of the border path.
```

gspencergoog commented on packages/flutter/lib/src/painting/borders.dart:
```diff
@@ -21,6 +21,27 @@ enum BorderStyle {
+  /// Border is drawn inside.
+  inside,
+
+  /// Border is drawn to the center, half the [BorderSide.width] to the inside,
+  /// the other half to the outside.
```
```suggestion
  /// The border is drawn on the center of the border path, with half of the
  ///  [BorderSide.width] on the inside, and the other half on the outside of the path.
```

gspencergoog commented on packages/flutter/lib/src/painting/borders.dart:
```diff
@@ -21,6 +21,27 @@ enum BorderStyle {
+  /// the other half to the outside.
+  center,
+
+  /// Border is drawn outside.
```
```suggestion
  /// The border is drawn on the outside of the border path.
```

gspencergoog commented on packages/flutter/lib/src/painting/borders.dart:
```diff
@@ -258,14 +296,15 @@ class BorderSide {
 
   @override
-  String toString() => '${objectRuntimeType(this, 'BorderSide')}($color, ${width.toStringAsFixed(1)}, $style)';
+  String toString() => '${objectRuntimeType(this, 'BorderSide')}($color, ${width.toStringAsFixed(1)}, $style, $strokeAlign)';
```
I think you can avoid printing the `, $strokeAlign` if it is set to `inside`.  This is consistent with how `Diagnosticable` handles default values, and would keep it from invalidating all the existing tests.

bernaferrari replied:
Fixed with an `if`, like `Diagnosticable`.

gspencergoog reviewed (changes requested):
Thanks for the PR, this is really a cool addition!

Just some minor things, I think it looks pretty good overall.

bernaferrari committed:
Review fixes.

bernaferrari committed:
Fix analyzer.

bernaferrari committed:
Add OutlinedBorder to Border documentation.

bernaferrari committed:
Fix stadium lerp.

bernaferrari commented:
Yay! I just fixed everything you mentioned and updated the sample for lerping between different Borders so we can see they are in a great state.

<img width="750" alt="image" src="https://user-images.githubusercontent.com/351125/169145892-45b7b554-793d-4e9e-87f1-51747860951b.png">

gspencergoog reviewed (approved):
![32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d](https://user-images.githubusercontent.com/8867023/37737237-66e9c9dc-2d10-11e8-8797-13a405eace49.png)

This is great!  Thanks!

gspencergoog commented:
Hmm. Looks like you got hit by a random CI issue where it didn't launch all the checks. To fix it, push an empty commit (i.e. use `git commit --allow-empty -m "Pushing empty commit"` and push) to try triggering it again.

bernaferrari commented:
It seems tree is broken by something else (all PR are broken). Maybe you should mark as "wait tree to go green".

gspencergoog commented:
Yes, the tree is _also_ broken at the moment, but because the "ci.yaml" step failed, you'll need to push an empty commit. That's the step that starts most of the CI checks.

bernaferrari commented:
I just pushed, but I think someone still didn't fix the ci.yaml issue. I can push again when they do.

bernaferrari commented:
rebasing to the top

gspencergoog commented:
Okay, looks like the checks fired off this time. And the tree is green, so once the tests pass, this will go in!

gspencergoog commented:
Ahh, sorry, I forgot we'll need a second review.  Let me see who can take a look.

darrenaustin commented on packages/flutter/lib/src/painting/box_border.dart:
```diff
@@ -411,6 +461,11 @@ class Border extends BoxBorder {
 
+  bool get _strokeAlignIsUniform {
+    final StrokeAlign topStrokeAlign = top.strokeAlign;
+    return right.strokeAlign == topStrokeAlign && bottom.strokeAlign == topStrokeAlign && left.strokeAlign == topStrokeAlign;
```
```suggestion
    return right.strokeAlign == topStrokeAlign
      && bottom.strokeAlign == topStrokeAlign
      && left.strokeAlign == topStrokeAlign;
```

darrenaustin commented on packages/flutter/lib/src/painting/box_border.dart:
```diff
@@ -688,9 +763,17 @@ class BorderDirectional extends BoxBorder {
 
+  bool get _strokeAlignIsUniform {
+    final StrokeAlign topStrokeAlign = top.strokeAlign;
+    return start.strokeAlign == topStrokeAlign && bottom.strokeAlign == topStrokeAlign && end.strokeAlign == topStrokeAlign;
```
```suggestion
    return start.strokeAlign == topStrokeAlign
      && bottom.strokeAlign == topStrokeAlign
      && end.strokeAlign == topStrokeAlign;
```

bernaferrari replied:
That also works. My original idea was to make it similar to the others 😅. Fixed.
<img width="1012" alt="image" src="https://user-images.githubusercontent.com/351125/169192354-8578650e-a3ed-4b91-ae1a-28104c95282c.png">

bernaferrari replied:
It is funny you prefer `&&` at the start than end. The file has both patterns, so it doesn't matter, I'll use exactly the way you wrote. But first time I see someone doing that.

gspencergoog replied:
We do it because with four spaces before the `&&`, and one after, it makes all of the terms line up under the return. It's not something we apply religiously, only in cases where there are a lot of conditions stacked up.

darrenaustin reviewed (approved):
LGTM as well. Just a couple of readability comments below. Nice work. Some tricky cases here.

bernaferrari committed:
Fix Darren review.

bernaferrari committed:
Final fixes.

- Fix BoxDecoration + BorderRadius(0) (which is different than BoxDecoration without a BorderRadius).
- Fix getInnerPath for CircleBorder.

bernaferrari committed:
Final fix 2.

Remove variable that may not always be needed.

bernaferrari committed:
Final fix 3.

Style fix.

bernaferrari committed:
Final fix final.

Add indentation.
